db/db_versioned.go

```diff
@@ -80,50 +96,192 @@ func (b *BoltDBVersioned) Stop(ctx context.Context) error {
+	if vn == nil {
+		// namespace not yet created
+		buf.Put(ns, _minKey, (&versionedNamespace{
+			keyLen: uint32(len(key)),
+		}).serialize(), "failed to create metadata")
+	} else {
+		if len(key) != int(vn.keyLen) {
+			return errors.Wrapf(ErrInvalid, "invalid key length, expecting %d, got %d", vn.keyLen, len(key))
+		}
+		last, _, err := b.get(math.MaxUint64, ns, key)
+		if !isNotExist(err) && version < last {
+			// not allowed to perform write on an earlier version
+			return ErrInvalid
+		}
+		buf.Delete(ns, keyForDelete(key, version), fmt.Sprintf("failed to delete key %x", key))
+	}
+	buf.Put(ns, keyForWrite(key, version), value, fmt.Sprintf("failed to put key %x", key))
+	return b.db.WriteBatch(buf)
 }
 
 // Get retrieves the most recent version
-func (b *BoltDBVersioned) Get(ns string, version uint64, key []byte) ([]byte, error) {
+func (b *BoltDBVersioned) Get(version uint64, ns string, key []byte) ([]byte, error) {
 	if !b.db.IsReady() {
 		return nil, ErrDBNotStarted
 	}
-	// TODO: implement Get
-	return nil, nil
+	// check key's metadata
+	if err := b.checkNamespaceAndKey(ns, key); err != nil {
+		return nil, err
+	}
+	_, v, err := b.get(version, ns, key)
+	return v, err
 }
 
-// Delete deletes a record,if key is nil,this will delete the whole bucket
-func (b *BoltDBVersioned) Delete(ns string, key []byte) error {
+func (b *BoltDBVersioned) get(version uint64, ns string, key []byte) (uint64, []byte, error) {
+	var (
+		last     uint64
+		isDelete bool
+		minimum  = keyForDelete(key, 0)
+		value    []byte
+	)
+	key = keyForWrite(key, version)
+	err := b.db.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(ns))
+		if bucket == nil {
+			return ErrBucketNotExist
+		}
+		c := bucket.Cursor()
+		k, v := c.Seek(key)
+		if k == nil || bytes.Compare(k, key) == 1 {
+			k, v = c.Prev()
+			if k == nil || bytes.Compare(k, minimum) <= 0 {
+				// cursor is at the beginning/end of the bucket or smaller than minimum key
+				return ErrNotExist
+			}
+		}
+		isDelete, last = parseKey(k)
+		value = make([]byte, len(v))
+		copy(value, v)
+		return nil
+	})
+	if err != nil {
+		return last, nil, err
+	}
+	if isDelete {
+		return last, nil, ErrDeleted
+	}
+	return last, value, nil
+}
+
+// Delete deletes a record, if key does not exist, it returns nil
+func (b *BoltDBVersioned) Delete(version uint64, ns string, key []byte) error {
 	if !b.db.IsReady() {
 		return ErrDBNotStarted
 	}
-	// TODO: implement Delete
-	return nil
+	// check key's metadata
+	if err := b.checkNamespaceAndKey(ns, key); err != nil {
+		return err
+	}
+	last, _, err := b.get(math.MaxUint64, ns, key)
+	if isNotExist(err) {
+		return err
+	}
+	if version < last {
+		// not allowed to perform delete on an earlier version
+		return ErrInvalid
+	}
+	buf := batch.NewBatch()
+	buf.Put(ns, keyForDelete(key, version), nil, fmt.Sprintf("failed to delete key %x", key))
+	buf.Delete(ns, keyForWrite(key, version), fmt.Sprintf("failed to delete key %x", key))
```

Comment:
why delete write key?

Reply:
discussed offline
1. it is to keep only the last operation on the key
2. i recall in the original POC, we have cases where the same key is deleted, then written in the batch at around height 400k. So it is a valid request (to delete then write on the same height). We'll allow this behavior (and also write then delete on the same height)